testresults/src/org/labkey/testresults/TestResultsController.java

```diff
@@ -273,52 +252,87 @@ public static void ensureRunDataCached(RunDetail[] runs, boolean keepObjData) {
+                    runMap.put("medianmem", medianmem);
+                    Table.update(null, TestResultsSchema.getTableInfoTestRuns(), runMap, run.getId());
 
                     // Set all to empty array so that memory does not run out.
                     // Each run ideally has 10k test passes so server
                     // could throw a mem error pretty quick if we didn't do this.
-                    if(!keepObjData) {
+                    if (!keepObjData) {
                         run.setPasses(new TestPassDetail[0]);
                         run.setFailures(new TestFailDetail[0]);
                         run.setTestmemoryleaks(new TestMemoryLeakDetail[0]);
+                        run.setHang(null);
                     }
                 }
             }
             transaction.commit();
         }
     }
 
-    public static User[] getTrainingDataForContainer(Container c) {
-        User[] users = new TableSelector(TestResultsSchema.getInstance().getTableInfoUser(), null, null).getArray(User.class);
+    public static User[] getTrainingDataForContainer(Container c, String username) {
+        SimpleFilter userFilter = null;
+        if (username != null && !username.isEmpty())
+        {
+            ArrayList<String> usernameCollection = new ArrayList<>();
+            usernameCollection.add(username);
+            userFilter = new SimpleFilter();
+            SimpleFilter.InClause in = new SimpleFilter.InClause(FieldKey.fromParts("username"), usernameCollection);
+            userFilter.addClause(in);
+        }
+
+        User[] users = new TableSelector(TestResultsSchema.getTableInfoUser(), userFilter, null).getArray(User.class);
         SQLFragment sqlFragment = new SQLFragment();
-        sqlFragment.append("SELECT * FROM testresults.userdata WHERE container = ?;");
-        sqlFragment.add(c.getEntityId().toString());
+        sqlFragment.append("SELECT * FROM testresults.userdata WHERE container = ?");
+        if (userFilter != null)
+        {
+            sqlFragment.append(" AND userid IN (");
+            for (int i = 0; i < users.length; i++)
+            {
+                if (i > 0)
+                    sqlFragment.append(',');
+                sqlFragment.append(users[i].getId());
+            }
+            sqlFragment.append(')');
+        }
+        sqlFragment.add(c.getEntityId());
         SqlSelector sqlSelector = new SqlSelector(TestResultsSchema.getSchema(), sqlFragment);
         sqlSelector.forEach(rs -> {
-            for(User user : users) {
-                if(user.getId() == rs.getInt("userid")) {
+            for (User user : users) {
+                if (user.getId() == rs.getInt("userid")) {
                     user.setMeanmemory(rs.getDouble("meanmemory"));
                     user.setMeantestsrun(rs.getDouble("meantestsrun"));
                     user.setStddevtestsrun(rs.getDouble("stddevtestsrun"));
```

Comment:
I did this, but I'm not sure how to make sure the SqlSelector is getting the right id column (since user/userdata both have an id column). Or if I need to?

Reply:
I'd switch from * to an explicit column list, and table-qualify the one you want. For example:

SELECT user.userid, test,meanmemory, test.meantestsrun FROM testresults.user user INNER JOIN testresults.userdata data ON user.id = data.userid WEHRE data.container = ? AND user.username=?